drivers/sds011/sds011.c

```diff
@@ -0,0 +1,351 @@
+/*
+ * Copyright (C) 2018 HAW-Hamburg
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     drivers_sds011
+ * @{
+ *
+ * @file
+ * @brief       SDS011 Laser Dust Sensor driver implementation
+ *
+ * @author      Michel Rottleuthner <michel.rottleuthner@haw-hamburg.de>
+ *
+ * @}
+ */
+
+#include <string.h>
+
+#include "assert.h"
+#include "sds011.h"
+#include "periph/uart.h"
+
+/** @brief UART receive callback */
+static void _rx_cb(void *arg, uint8_t data)
+{
+    sds011_t *dev = (sds011_t*)arg;
+
+    /* frame MUST start with HEAD byte and the buffer must be cleared
+       before writing to it again */
+    if (((dev->pos == 0) && (data != SDS011_FRAME_HEAD)) ||
+         (dev->pos == SDS011_FRAME_RECV_LEN)) {
+        return;
+    }
+
+    dev->rx_mem[dev->pos] = data;
+
+    if ((dev->pos >= SDS011_DB1_IDX) &&
+        (dev->pos < SDS011_FRAME_RECV_CSUM_IDX)) {
+        dev->checksum += data;
+    }
+    else if ((dev->pos == SDS011_FRAME_RECV_LEN - 1) &&
+             (dev->rx_mem[SDS011_FRAME_HEAD_IDX] == SDS011_FRAME_HEAD) &&
+             (dev->rx_mem[SDS011_FRAME_RECV_TAIL_IDX] == SDS011_FRAME_TAIL)) {
+
+        dev->checksum &= SDS011_FRAME_CSUM_MSK;
+        if (dev->rx_mem[SDS011_FRAME_RECV_CSUM_IDX] == dev->checksum) {
+
+            if ((dev->cb != NULL) &&
+                (dev->rx_mem[SDS011_CMDID_IDX] == SDS011_RCMDID_DATA)) {
+                    sds011_data_t measure;
+                    measure.pm_2_5 = dev->rx_mem[SDS011_DB1_IDX] |
+                                    (dev->rx_mem[SDS011_DB2_IDX] << 8);
+                    measure.pm_10  = dev->rx_mem[SDS011_DB3_IDX] |
+                                    (dev->rx_mem[SDS011_DB4_IDX] << 8);
+                    dev->cb(&measure, dev->cbctx);
+                    dev->pos = -1;
+            }
+        }
+        else {
+            dev->pos = -1;
+        }
+
+        dev->checksum = 0;
+
+        /* unlock the mutex for the calling function */
+        mutex_unlock(&dev->cb_lock);
+    }
+
+    dev->pos++;
+}
+
+/** @brief send command and wait for the first message that is sent as reply */
+int _send_recv_cmd(sds011_t *dev, uint8_t *data_bytes, size_t len, uint8_t *recv_frm)
+{
+    uint8_t cmd[SDS011_FRAME_SEND_LEN] = {0};
+    int checksum = 0;
+    int res = 0;
```

Comment:
should initialised with proper `enum` value, e.g. `SDS011_ERROR` as default

Reply:
done